Fixes #977: Adding reference pitch to tuning conversion utils

jlw365 opened the pull request:
#### Reference Issue
Fixes #977 

#### What does this implement/fix? Explain your changes.
- Added `A4_to_tuning` and `tuning_to_A4` conversion functions to allow someone to tune to/work with something besides A440 if desired.
- Added associated unit tests
- In the future, an additional argument could be added to functions such as `midi_to_hz` utilizing these conversion functions, but that's outside of this scope for now.

#### Any other comments?
A few open questions:
- Does it make sense to have these take a `float` or `float or np.ndarray` A4/tuning? Currently wrote them to match the majority of the other unit conversions being  with `float or np.ndarray`.
- In the docstrings, I rounded the example usage results to the thousandths, but there are some small inconsistencies with that in the tests [here ](https://github.com/librosa/librosa/compare/master...jlw365:tuning-conversion-util?expand=1#diff-212b29ab51fbabe0781c6aac7dd4124fR119) - not sure how to handle that.
- I considered adding defaults of `A=440.0` and `tuning=0.0`, but not sure if that makes sense since you wouldn't come to use this anyways if you're doing that?
- Let me know if this wasn't what you had in mind or if I should format any parts here differently; I wasn't super confident in the best style of unit tests that would make sense for these methods.

Thanks!

bmcfee commented:
Thanks! This looks great so far.  Don't worry about the failing travis tests; I'm fixing that in a separate PR.  (I might ask you to rebase before we merge though.)

Per questions:


>     * Does it make sense to have these take a `float` or `float or np.ndarray` A4/tuning? Currently wrote them to match the majority of the other unit conversions being  with `float or np.ndarray`.

Matching existing functions seems good to me.



>     * In the docstrings, I rounded the example usage results to the thousandths, but there are some small inconsistencies with that in the tests [here ](https://github.com/librosa/librosa/compare/master...jlw365:tuning-conversion-util?expand=1#diff-212b29ab51fbabe0781c6aac7dd4124fR119) - not sure how to handle that.

That's fine.  I usually modify numpy's default print settings for docstring examples just to make them easier to read and format cleanly.  We don't run doctests (but maybe should start?), so it's not a huge deal.



>     * I considered adding defaults of `A=440.0` and `tuning=0.0`, but not sure if that makes sense since you wouldn't come to use this anyways if you're doing that?

I think we can require these arguments (not provide defaults).  However, the docstring examples seem to run with defaults, so we should be consistent about it.



>     * Let me know if this wasn't what you had in mind or if I should format any parts here differently; I wasn't super confident in the best style of unit tests that would make sense for these methods.

Seems great!  I have some comments on the docs, but will leave those in CR.  Otherwise, tests seem fine for what this is, and I don't think we need to go overboard with error handling or anything.

bmcfee commented on librosa/core/time_frequency.py:
```diff
@@ -900,6 +902,70 @@ def octs_to_hz(octs, tuning=0.0, bins_per_octave=12):
+    >>> librosa.A4_to_tuning(432.0)
+    -0.318
+    >>> librosa.A4_to_tuning([440.0, 444.0], 24)
+    array([   0.,   0.313   ])
```
In the examples, it's helpful to have a comment before each one explaining what it does.  [This one](https://librosa.github.io/librosa/generated/librosa.core.tone.html#librosa.core.tone) gives an example of what I mean.

bmcfee commented on librosa/core/time_frequency.py:
```diff
@@ -900,6 +902,70 @@ def octs_to_hz(octs, tuning=0.0, bins_per_octave=12):
+    Examples
+    --------
+    >>> librosa.A4_to_tuning()
+    0.
```
This example won't actually run because it uses a default value.

bmcfee commented on librosa/core/time_frequency.py:
```diff
@@ -900,6 +902,70 @@ def octs_to_hz(octs, tuning=0.0, bins_per_octave=12):
 
 
+def A4_to_tuning(A, bins_per_octave=12):
+    """Convert reference frequencies to cents (bin fractions).
```
Docstring could be more informative here about what reference frequency means, and how this function should be used.

Maybe something more like:
```
Convert a reference pitch frequency (e.g., `A4=435`) to a tuning estimation, in fractions of a bin per octave.

This is useful for determining the tuning deviation relative to A440 of a given frequency, assuming equal temperament.  By default, 12 bins per octave are used.
```

Note the output wouldn't really be cents, but 1/100ths of a cent.

bmcfee commented on librosa/core/time_frequency.py:
```diff
@@ -900,6 +902,70 @@ def octs_to_hz(octs, tuning=0.0, bins_per_octave=12):
+    return bins_per_octave * (np.log2(np.asanyarray(A)) - np.log2(440.0))
+
+
+def tuning_to_A4(tuning, bins_per_octave=12):
```
Same comments as above.

bmcfee commented on librosa/core/time_frequency.py:
```diff
@@ -900,6 +902,70 @@ def octs_to_hz(octs, tuning=0.0, bins_per_octave=12):
     return (float(A440) / 16)*(2.0**np.asanyarray(octs))
 
 
+def A4_to_tuning(A, bins_per_octave=12):
```
Question for the peanut gallery: do we want this to be `A` or `A4`?  Neither is great python style, but I think the latter is more precise.  It's a positional argument, so it doesn't really matter that much, but I'm curious for people's thoughts.

bmcfee commented:
Also, I've merged the fix for the travis builds.  If you could rebase this PR to current master, the tests should run cleanly again.

jlw365 commented:
Addressed changes above, adding more descriptive docstrings and example descriptions, fixing the bug in the example code using defaults, and changing the parameter in `A4_to_tuning` to be `A4` instead of `A`. Rebased to address the CI issues, but still working on some issues with the rebase/squashing above. Thanks for the helpful feedback!

bmcfee commented:
@jlw365 thanks again for this!  Tests are passing and all looks good to me.  Anything else you'd like to add, or shall I :shipit: ?

lostanlen commented:
Thanks for all this, @jlw365 

Part of me thinks that this subject (tuning and midi-hz conversions) would be worthy of a blog post
https://librosa.github.io/blog/

bmcfee commented:
> Part of me thinks that this subject (tuning and midi-hz conversions) would be worthy of a blog post
> https://librosa.github.io/blog/

Indeed -- want to put it on the todo list for the blog?

jlw365 commented:
> @jlw365 thanks again for this! Tests are passing and all looks good to me. Anything else you'd like to add, or shall I :shipit: ?

Thanks! Sounds good to me 🚀 🚢

bmcfee commented:
Thanks again @jlw365 !
